File: tests/test_anomaly_benchmark.py
import os
import sys
import json
import pytest
import numpy as np
import pandas as pd

BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if BASE_DIR not in sys.path:
    sys.path.insert(0, BASE_DIR)

from src.anomaly_detection.robust_mad import RobustMADDetector
from src.anomaly_detection.copod import COPODDetector
from src.anomaly_detection.isolation_forest import IsolationForestDetector, euler_harmonic_c

BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
DATASET_PATH = os.path.join(BASE_DIR, "data", "synthetic", "semiconductor_synthetic_full.csv")
SPLIT_MANIFEST_PATH = os.path.join(BASE_DIR, "ml", "data", "split_manifest.json")
ANOMALY_CONTRACT_PATH = os.path.join(BASE_DIR, "ml", "anomaly", "anomaly_contract.json")
PROD_ARTIFACT_V2_PATH = os.path.join(BASE_DIR, "ml", "models", "production", "predicta_anomaly_v2_artifacts.json")


@pytest.fixture
def sample_train_data():
    np.random.seed(42)
    n = 100
    df = pd.DataFrame({
        "iddq": np.random.normal(2100.0, 50.0, n),
        "ileak": np.random.normal(300.0, 10.0, n),
        "tpd": np.random.normal(190.0, 5.0, n),
    })
    lots = pd.Series(["LOT-SYN-001"] * 50 + ["LOT-SYN-002"] * 50)
    return df, lots


def test_anomaly_contract_structure():
    """Verify anomaly contract exists, is valid JSON, and adheres to schema."""
    assert os.path.exists(ANOMALY_CONTRACT_PATH)
    with open(ANOMALY_CONTRACT_PATH, "r", encoding="utf-8") as f:
        contract = json.load(f)

    assert contract["contract_version"] == "2.0.0"
    assert contract["data_governance"]["is_synthetic"] is True
    assert "e2b969c458864b11ed61a6073ed1356adcbfd6775bb2c44b28023446bf9771fa" in contract["data_governance"]["dataset_sha256"]
    assert contract["detector_definitions"]["isolation_forest"]["random_state"] == 42
    assert contract["detector_definitions"]["robust_mad"]["min_reference_size"] == 10


def test_deterministic_isolation_forest(sample_train_data):
    """Verify Isolation Forest produces identical trees across runs with seed=42."""
    df, lots = sample_train_data
    clf1 = IsolationForestDetector(n_estimators=20, random_state=42)
    clf1.fit(df, lots)

    clf2 = IsolationForestDetector(n_estimators=20, random_state=42)
    clf2.fit(df, lots)

    s1 = clf1.score(df)
    s2 = clf2.score(df)
    np.testing.assert_allclose(s1, s2, atol=1e-7)


def test_feature_order_locking(sample_train_data):
    """Verify Isolation Forest locks canonical feature order."""
    df, lots = sample_train_data
    clf = IsolationForestDetector(n_estimators=10, random_state=42)
    clf.fit(df, lots)
    assert clf.feature_names == ["iddq", "ileak", "tpd"]

    # Reordered input raises or matches correctly
    reordered_df = df[["tpd", "iddq", "ileak"]].copy()
    clf.fit(reordered_df, lots)
    assert clf.feature_names == ["tpd", "iddq", "ileak"]


def test_robust_mad_lot_relative_and_fallback(sample_train_data):
    """Verify Robust MAD uses lot stats for known lots and falls back safely for unseen/small lots."""
    df, lots = sample_train_data
    det = RobustMADDetector(min_reference_size=10)
    det.fit(df, lots)

    # 1. Known Lot
    res_known = det.score_single({"iddq": 2100.0, "ileak": 300.0, "tpd": 190.0}, "LOT-SYN-001")
    assert res_known["reference_source"] == "LOT_RELATIVE"

    # 2. Unseen Lot
    res_unseen = det.score_single({"iddq": 2100.0, "ileak": 300.0, "tpd": 190.0}, "LOT-UNSEEN-999")
    assert "GLOBAL_FALLBACK" in res_unseen["reference_source"]

    # 3. Missing Lot
    res_missing = det.score_single({"iddq": 2100.0, "ileak": 300.0, "tpd": 190.0}, None)
    assert res_missing["reference_source"] == "GLOBAL_FALLBACK"


def test_copod_monotonicity():
    """Verify COPOD tail probability increases with extremity."""
    ref_df = pd.DataFrame({
        "iddq": np.linspace(1000.0, 2000.0, 100),
        "ileak": np.linspace(100.0, 200.0, 100),
        "tpd": np.linspace(10.0, 20.0, 100),
    })
    copod = COPODDetector()
    copod.fit(ref_df)

    s_median = copod.score_single({"iddq": 1500.0, "ileak": 150.0, "tpd": 15.0})["score"]
    s_extreme = copod.score_single({"iddq": 2500.0, "ileak": 350.0, "tpd": 35.0})["score"]
    assert s_extreme > s_median


def test_insufficient_reference_small_lot():
    """Verify lots with sample count < min_reference_size do not form lot-specific models."""
    df = pd.DataFrame({
        "iddq": [2100.0] * 5,
        "ileak": [300.0] * 5,
        "tpd": [190.0] * 5,
    })
    lots = pd.Series(["TINY-LOT"] * 5)
    det = RobustMADDetector(min_reference_size=10)
    det.fit(df, lots)
    assert "TINY-LOT" not in det.lot_stats
    res = det.score_single({"iddq": 2100.0, "ileak": 300.0, "tpd": 190.0}, "TINY-LOT")
    assert "GLOBAL_FALLBACK" in res["reference_source"]


def test_nan_inf_rejection():
    """Verify detectors reject or handle NaN/Inf cleanly."""
    det = RobustMADDetector()
    det.global_stats = {
        "iddq": {"median": 2000.0, "sigma": 100.0},
        "ileak": {"median": 300.0, "sigma": 10.0},
        "tpd": {"median": 190.0, "sigma": 5.0},
    }
    res = det.score_single({"iddq": np.nan, "ileak": 300.0, "tpd": 190.0})
    assert "iddq" not in res["parameter_z_scores"]


def test_v2_production_artifact_exists_and_valid():
    """Verify predicta_anomaly_v2_artifacts.json exists, contains all required fields, and passes hash check."""
    assert os.path.exists(PROD_ARTIFACT_V2_PATH)
    with open(PROD_ARTIFACT_V2_PATH, "r", encoding="utf-8") as f:
        artifact = json.load(f)

    assert artifact["contract_version"] == "2.0.0_authoritative"
    assert "robust_mad" in artifact
    assert "copod" in artifact
    assert "isolation_forest" in artifact
    assert "fusion" in artifact
    assert "frozen_thresholds" in artifact
    assert "trees" in artifact["isolation_forest"]
    assert len(artifact["isolation_forest"]["trees"]) == 100


def test_euler_harmonic_math():
    """Verify average BST path length calculation matches theory."""
    assert euler_harmonic_c(1) == 0.0
    assert euler_harmonic_c(2) == 1.0
    c_256 = euler_harmonic_c(256)
    assert 9.0 < c_256 < 11.0

File: tests/test_anomaly_benchmark.js
/**
 * Predicta Semiconductor Intelligence Platform — Anomaly Benchmark & Parity Tests (Node.js)
 * File: tests/test_anomaly_benchmark.js
 */

const fs = require('fs');
const path = require('path');
const { RobustMADDetectorJS } = require('../src/anomaly_detection/robust_mad');
const { COPODDetectorJS } = require('../src/anomaly_detection/copod');
const { IsolationForestDetectorJS, eulerHarmonicC } = require('../src/anomaly_detection/isolation_forest');
const { AnomalyFusionEngineJS } = require('../src/anomaly_detection/fusion');

console.log("Starting Stage 4 Dynamic Anomaly Benchmark Unit & Parity Tests...\n");

let failures = 0;

function assert(condition, message) {
  if (condition) {
    console.log(`[PASS] ${message}`);
  } else {
    console.error(`[FAIL] ${message}`);
    failures++;
  }
}

// 1. Contract & Artifact Existence
const artifactPath = path.join(__dirname, '../ml/models/production/predicta_anomaly_v2_artifacts.json');
const contractPath = path.join(__dirname, '../ml/anomaly/anomaly_contract.json');
const reportJsonPath = path.join(__dirname, '../experiments/anomaly_evaluation/anomaly_benchmark_report.json');
const reportMdPath = path.join(__dirname, '../experiments/anomaly_evaluation/anomaly_benchmark_report.md');

assert(fs.existsSync(artifactPath), "Production V2 anomaly artifact exists (predicta_anomaly_v2_artifacts.json)");
assert(fs.existsSync(contractPath), "Authoritative anomaly contract exists (anomaly_contract.json)");
assert(fs.existsSync(reportJsonPath), "Benchmark JSON report exists (anomaly_benchmark_report.json)");
assert(fs.existsSync(reportMdPath), "Benchmark Markdown report exists (anomaly_benchmark_report.md)");

const artifact = JSON.parse(fs.readFileSync(artifactPath, 'utf-8'));
assert(artifact.contract_version === "2.0.0_authoritative", "V2 artifact contract_version is '2.0.0_authoritative'");
assert(artifact.threshold_optimization_metric === "F2_MAX_VALIDATION_ONLY", "Artifact records F2_MAX_VALIDATION_ONLY optimization metric");
assert(Array.isArray(artifact.canonical_feature_order) && artifact.canonical_feature_order.length === 3, "Canonical features contain exactly 3 dimensions");
assert(artifact.canonical_feature_order[0] === "iddq" && artifact.canonical_feature_order[1] === "ileak" && artifact.canonical_feature_order[2] === "tpd", "Canonical feature order is exactly [iddq, ileak, tpd]");
assert(artifact.isolation_forest && Array.isArray(artifact.isolation_forest.trees) && artifact.isolation_forest.trees.length === 100, "Isolation Forest contains 100 serialized trees");

// 2. Mathematical Parity & Tree Traversal Test
const isoDet = new IsolationForestDetectorJS(artifact.isolation_forest);
const testNormalSample = { iddq: 2140.0, ileak: 300.0, tpd: 192.0 };
const testAnomalousSample = { iddq: 4500.0, ileak: 850.0, tpd: 310.0 };

const resNormal = isoDet.scoreSingle(testNormalSample);
const resAnomaly = isoDet.scoreSingle(testAnomalousSample);

assert(resNormal.score < resAnomaly.score, `Isolation forest scores anomaly higher than normal: Normal=${resNormal.score}, Anomaly=${resAnomaly.score}`);
assert(resAnomaly.score > 0.60, `Extreme outlier produces high anomaly score: ${resAnomaly.score}`);
assert(typeof resAnomaly.anomaly_evidence === 'object', "Isolation forest produces feature-level anomaly evidence");

// 3. Schema & Feature Order Locking in Node.js
let missingThrew = false;
try {
  isoDet.scoreSingle({ iddq: 2100.0, ileak: 300.0 });
} catch (e) {
  missingThrew = true;
}
assert(missingThrew, "Isolation Forest rejects missing feature with explicit error");

let nonNumericThrew = false;
try {
  const madDetTemp = new RobustMADDetectorJS(artifact.robust_mad);
  madDetTemp.scoreSingle({ iddq: "invalid_string", ileak: 300.0, tpd: 190.0 });
} catch (e) {
  nonNumericThrew = true;
}
assert(nonNumericThrew, "Robust MAD rejects non-numeric value with explicit error");

// 4. Robust MAD Lot-Relative vs Fallback
const madDet = new RobustMADDetectorJS(artifact.robust_mad);
const resKnown = madDet.scoreSingle(testNormalSample, "LOT-SYN-001");
const resUnseen = madDet.scoreSingle(testNormalSample, "LOT-SYN-999");
const resMissing = madDet.scoreSingle(testNormalSample, null);

assert(resKnown.reference_source === "LOT_RELATIVE", `Known lot uses LOT_RELATIVE statistics (found: ${resKnown.reference_source})`);
assert(resUnseen.reference_source.includes("GLOBAL_FALLBACK"), `Unseen lot falls back to global baseline (found: ${resUnseen.reference_source})`);
assert(resMissing.reference_source === "GLOBAL_FALLBACK", `Missing lot falls back to global baseline (found: ${resMissing.reference_source})`);

// 5. Multi-Criteria Fusion Engine
const fusionEngine = new AnomalyFusionEngineJS({
  mad_parameters: artifact.robust_mad,
  copod_parameters: artifact.copod,
  isolation_forest_parameters: artifact.isolation_forest,
  weights: artifact.fusion.weights,
  normalization_scales: artifact.fusion.normalization_scales,
  fusion_threshold: artifact.fusion.fusion_threshold,
});

const fusionNorm = fusionEngine.evaluateComponent(testNormalSample, "LOT-SYN-001");
const fusionAnom = fusionEngine.evaluateComponent(testAnomalousSample, "LOT-SYN-001");

assert(fusionNorm.overall_status === "PASS", `Normal sample evaluated as PASS by fusion engine (found: ${fusionNorm.overall_status})`);
assert(fusionAnom.overall_status === "REJECT", `Anomalous sample evaluated as REJECT by fusion engine (found: ${fusionAnom.overall_status})`);
assert(fusionAnom.conservative_alarm === true, "Conservative alarm triggered for extreme anomaly");

// 6. BST Harmonic Math Test
assert(eulerHarmonicC(1) === 0.0, "BST c(1) is 0");
assert(eulerHarmonicC(2) === 1.0, "BST c(2) is 1");
const c256 = eulerHarmonicC(256);
assert(c256 > 9.0 && c256 < 11.0, `BST c(256) ≈ 10.24 (found: ${c256.toFixed(4)})`);

console.log(`\nTests completed with ${failures} failures.`);

if (failures > 0) {
  process.exit(1);
} else {
  console.log("All Stage 4 Dynamic Anomaly Benchmark tests passed successfully!");
  process.exit(0);
}
